### PR TITLE
fix: check for invalid arguments in `getScores`

### LINKED ISSUE #923: fix: add more validation rules to `getScores`

### PR DESCRIPTION
**Pull request by wa0x6e**
## 🧿 Current issues / What's wrong ?

From https://github.com/orgs/snapshot-labs/projects/12/views/7?pane=issue&itemId=43073796

The `getScores` passes the arguments as is to score-api without check. There are some basic check possible from the SDK, to filter out invalid requests, and lighten the score-api server load.

This will also avoid log pollution on score-api, making the log more relevant

## 💊 Fixes / Solution

Fix #923  

Add some basic argument validation on the SDK side, and return a meaningful error when argument are not valid.

## 🚧 Changes

- Add check that `addresses` should always be an array, and contains always at least on valid address
- Add check that network is not empty, and valid (referenced in networks.json)
- Check that the given `snapshot` is in the network range (should be greater than the network's `start` block)
- Check that `network` in strategies are valid
- Fully tests `getScores`

## 🛠️ Tests

- Send some request to `getScores` functions, with invalid data
- The function should throw an error with an error message

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -222,6 +223,33 @@ export async function getScores(
...
+  }
+  strategies.forEach((strategy) => {
+    if (strategy.network && !networks[strategy.network]) {
+      throw new Error(
+        `Invalid network (${strategy.network}) in strategy ${strategy.name}`
+      );
+    }
+  });
+  if (typeof snapshot === 'number' && snapshot < networks[network].start) {
+    throw new Error(
```
[the remark this answers is not in the document]
Comment (wa0x6e) [reply src/utils.ts:248]: Correct

File: src/utils.ts [hunk 2]
```diff
@@ -222,6 +223,37 @@ export async function getScores(
...
+          `Invalid network (${strategy.network}) in strategy ${strategy.name}`
+        );
+      }
+    });
+  } catch (e) {
+    return Promise.reject(e);
+  }
+  if (!isValidSnapshot(snapshot, network)) {
+    return Promise.reject(
+      `Snapshot (${snapshot}) must be greater than network start block (${networks[network].start})`
```
Comment (ChaituVR) [inline src/utils.ts:253]:
```suggestion
      `Snapshot (${snapshot}) must be 'latest' or greater than network start block (${networks[network].start})`
```

File: src/utils.ts [hunk 3]
```diff
@@ -497,6 +529,21 @@ export function getNumberWithOrdinal(n) {
...
+}
+
+function isValidAddress(address: string) {
+  return isAddress(address) && address !== EMPTY_ADDRESS;
+}
+
+function isValidSnapshot(snapshot: number | string, network: string) {
+  return (
+    (typeof snapshot === 'number' && snapshot >= networks[network].start) ||
+    snapshot === 'latest'
```
Comment (ChaituVR) [inline src/utils.ts:543]:
```suggestion
    snapshot === 'latest' ||
    (typeof snapshot === 'number' && snapshot >= networks[network].start)
```

File: src/utils.spec.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+          'addresses is an empty array',
+          { addresses: [] },
+          /addresses can not be empty/i
+        ],
+        ['network is not valid', { network: 'mainnet' }, /invalid network/i],
+        ['network is empty', { network: '' }, /invalid network/i],
+        [
+          'snapshot is smaller than start block',
+          { snapshot: 1234 },
+          /snapshot \([0-9]+\) must be greater than network start block/i
```
Comment (ChaituVR) [inline src/utils.spec.js:82]:
```suggestion
          /snapshot \([0-9]+\) must be 'latest' or greater than network start block/i
```

File: src/utils.ts [hunk 5]
```diff
@@ -249,6 +250,37 @@ export async function getScores(
...
+    return Promise.reject('addresses should be an array of addresses');
+  }
+  if (addresses.length === 0) {
+    return Promise.reject('addresses can not be empty');
+  }
+  for (const address of addresses) {
+    if (!isValidAddress(address)) {
+      return Promise.reject(`Invalid address: ${address}`);
+    }
+  }
```
Comment (ChaituVR) [inline src/utils.ts:263]: I feel like better to use `.some`

Comment (ChaituVR) [reply src/utils.ts:263]:
```suggestion
  if (addresses.some(address => !isValidAddress(address))) {
    return Promise.reject(`Invalid address: ${address}`);
  }
```

File: src/utils.ts [hunk 6]
```diff
@@ -249,6 +250,37 @@ export async function getScores(
...
+    strategies.forEach((strategy) => {
+      if (strategy.network && !isValidNetwork(strategy.network)) {
+        throw new Error(
+          `Invalid network (${strategy.network}) in strategy ${strategy.name}`
+        );
+      }
+    });
+  } catch (e) {
+    return Promise.reject(e);
+  }
```
Comment (ChaituVR) [inline src/utils.ts:277]: Why different approach here (try,catch), can be the same `Promise.reject` like every where else

Comment (wa0x6e) [reply src/utils.ts:277]: Because it can not reject inside forEach because of the anonymous function. Will use `.some` as suggested above

Comment (ChaituVR) [reply src/utils.ts:275]:
```suggestion
  const invalidStrategy = strategies.find(strategy => strategy.network && !isValidNetwork(strategy.network));
  
  if (invalidStrategy) {
    return Promise.reject(new Error(`Invalid network (${invalidStrategy.network}) in strategy ${invalidStrategy.name}`));
  }
```

**Review by ChaituVR — APPROVED** [review]
tACK

**Pull request merged**